### ISSUE TITLE
Final hail mary

### ISSUE THREAD

**Issue by athariandre** [athariandre/yasta#9]
# MEGA-PR SPEC — Tron PPO Agent v1 (All‑in‑One)

**Goal**: Land a single, production‑ready PR that trains a self‑play Tron agent with PPO on CPU. The PR must be reproducible, numerically stable, and easy to monitor/debug.

---

## 0) Outcomes & Definition of Done

* ✅ PPO training runs end‑to‑end on CPU and improves vs. a frozen copy baseline.
* ✅ Self‑play with a frozen opponent policy is stable; periodic syncs do not destabilize learning.
* ✅ Checkpoints and logs are saved; runs are reproducible with a fixed seed.
* ✅ Action masking prevents illegal moves at both rollout and update time.
* ✅ GAE/returns are correct across episode boundaries (no bootstrap leakage).
* ✅ All tensors and buffers live on the configured device (`cpu`).
* ✅ NaN/inf guards exist for logits, probabilities, advantages, losses, and KL.
* ✅ Unit + integration tests pass locally (CPU only) and within Docker.
* ✅ Performance budgets: <5ms forward per step (policy), <5MB model size.

---

### Existing

* `training/policy.py` — Actor‑Critic network + action masking + evaluate_actions.
* `training/rollout.py` — Rollout collector + GAE + batching.
* `training/trainer.py` — PPO loop, checkpointing, logging.
* `training/config.py` — Hyperparameters and constants.
* `training/env.py` — TronEnv (provides obs + step + result_pov).

### New/Updated

* `training/utils/logging.py` — CSV/TB logging helpers (optional TB).
* `scripts/run_training.sh` — One‑shot launcher (reproducible flags).
* `Dockerfile` — CPU image with pinned deps.
* `pytest.ini` — Test config.
* `tests/` — Unit + integration tests (see §10).
* `README.md` (training section) — How to run + interpret logs.

---

### Observation dict (from `TronEnv`)

* `board: np.float32[H,W]` values in {0,1,2}, 0=empty, non‑zero occupied.
* `my_head: np.float32[2]` (x, y) grid coords (0‑based, integers cast as floats).
* `opp_head: np.float32[2]` (x, y) grid coords.
* `boosts: np.float32[2]` remaining boost counters for [me, opp].
* `turns: np.float32[1]` current turn index.
* `directions: np.float32[2]` encoded headings ∈ {0,1,2,3}.
* `alive: np.float32[2]` binary ∈ {0,1} for [me, opp].

### Action space

* Discrete {0: up, 1: right, 2: down, 3: left}.

### Mask

* `np.float32[num_actions]` with 1.0 for legal, 0.0 for illegal.
* Guarantee: if all zero → fallback to all ones (safeguard).

---

## 3) Config Additions & Defaults (`training/config.py`)

```python
SEED = 42
DEVICE = "cpu"
BOARD_H = 15; BOARD_W = 15
NUM_ACTIONS = 4
LATENT_SIZE = 128  # Keep model <5MB
MAX_TRAINING_STEPS = 200_000
ROLLOUT_LENGTH = 2048
BATCH_SIZE = 256
NUM_EPOCHS = 10
LEARNING_RATE = 3e-4
GAMMA = 0.99
GAE_LAMBDA = 0.95
PPO_CLIP_EPS = 0.2
VALUE_CLIP_RANGE = 0.2
VALUE_COEFF = 0.5
ENTROPY_COEFF = 0.01
MAX_GRAD_NORM = 0.5
MAX_KL = 0.03
LOG_INTERVAL = 4096
CHECKPOINT_INTERVAL = 50_000
USE_FROZEN_OPPONENT = True
OPPONENT_UPDATE_INTERVAL = 5  # in updates, not steps
MAX_BOOST = 3
MAX_TURNS = 1000
RUN_NAME = None  # auto generated if None
LOG_DIR = "runs/"
CKPT_DIR = "training/checkpoints/"
CSV_LOG = True
TENSORBOARD = False  # optional
```

Requirements:

* All hyperparams configurable via CLI or env vars (see §9).
* `RUN_NAME` default: `tronPPO_{YYYYmmdd_HHMMSS}_seed{SEED}`.

---

### Constraints

* CPU‑only (`.to(device)` with `device=cpu`).
* 2‑layer shared MLP encoder → actor/critic MLP heads (already present).
* Model size <5MB (print warning if exceeded).
* Forward latency budget <5ms on typical CPU (documented in README).

### Input Encoding

* Flatten board to `H*W`, divide by 2.0 → [0,1].
* Normalize heads by `[W,H]`.
* Normalize `boosts / MAX_BOOST`, `turns / MAX_TURNS`, `directions / 3.0`.
* `alive` left as {0,1}.
* Concatenate in fixed order; assert final length `feature_size`.

### Action Masking

* Compute from current head and next‑cell occupancy; block out-of-bounds and occupied.
* At `act()` and `evaluate_actions()`, replace invalid logits with large negative constant (−1e8).
* Guarantee device consistency (`mask_tensor.to(self.device)`).

### Numerical Safety

* After softmax, clamp probs to `[1e-8, 1.0]` via `torch.clamp_min` if needed before log.
* In `evaluate_actions`, ensure `actions` shape is `(B,)` and dtype `long`.
* KL estimate: use `old_logprobs - new_logprobs`, take mean of absolute value; if non‑finite, return large sentinel to trigger early stop.

### API

* `encode_obs(obs) -> torch.FloatTensor[feature_size] (detached, on device)`
* `act(obs, action_mask=None) -> (action:int, logprob:float, value:float)`
* `evaluate_actions(obs_tensors, actions, action_masks=None, old_values=None, old_logprobs=None) -> (logprobs, values, entropy, kl)`
* `get_value(obs) -> float`

---

### Collection

* For each step: compute mask → `policy.act(obs, mask)` → `env.step(action)`.
* Store raw obs dicts; actions, masks, logprobs, values, rewards, dones.
* Track `result_pov` in `episode_info.completed_episodes[i].outcome`.

### GAE/Returns

* Use `dones[t]` to stop bootstrapping at the boundary (no leakage to next episode). Last step uses `bootstrap_value` if `final_done=False`.
* Implement exactly:

  * `delta_t = r_t + gamma * V_{t+1} * (1 - done_t) - V_t`
  * `A_t = delta_t + gamma * lambda * (1 - done_t) * A_{t+1}`
  * `returns = advantages + values`
* Unit test covers terminal vs. non‑terminal rollouts (see §10.2).

### Batch Prep

* Convert arrays → tensors on `config.DEVICE`.
* Normalize advantages (z‑score with `+1e-8`).
* `get_mini_batches`: shuffle indices; encode obs lazily via `policy.encode_obs` to ensure consistent normalization.

### Safety

* Assert shapes: lengths of all arrays equal; masks shape `(T, NUM_ACTIONS)`.
* If `mask.sum()==0` at any step, replace mask with all ones and log a warning counter.

---

### Seed & Reproducibility

* `set_seed(SEED)` sets Python, NumPy, Torch CPU seeds.
* Print seed and run name at start.

### Self‑Play Frozen Opponent

* If `USE_FROZEN_OPPONENT=True`: clone weights initially; set `requires_grad=False` for opponent.
* After every `OPPONENT_UPDATE_INTERVAL` updates, copy main policy weights to opponent and log `[Opponent updated at step X]`.

### PPO Update Loop

For each training iteration:

1. **Rollout**: `T = ROLLOUT_LENGTH` steps.
2. **GAE/Returns**: compute with bootstrap.
3. **Prepare Batch**: tensors to device, normalize adv.
4. **Epochs**: `NUM_EPOCHS` times:

   * Iterate shuffled mini‑batches.
   * `policy.evaluate_actions(...)` to get `(logprobs, values, entropy, kl)`.
   * `ratio = exp(logprobs - old_logprobs)`.
   * Policy loss: `-min(ratio*adv, clamp(ratio)*adv)`.
   * Value loss: max(MSE(values, returns), MSE(clipped_values, returns)).
   * Entropy bonus: subtract `ENTROPY_COEFF * entropy.mean()`.
   * Backprop: zero‑grad → backward → grad‑clip (`MAX_GRAD_NORM`) → step.
   * Metrics: policy/value loss, entropy, clip frac, KL.
5. **Early Stop** if mean KL in epoch > `MAX_KL`.
6. **Log** every `LOG_INTERVAL` steps: win/draw rates, avg reward/len, times.
7. **Checkpoint** when `total_steps` crosses multiples of `CHECKPOINT_INTERVAL`.

### NaN/Inf Guards

* If any tensor used in loss is non‑finite → skip step and log `WARNING: non‑finite ...` with counts.
* If KL non‑finite → set to `MAX_KL*2` so epoch early‑stops.

### Checkpoint Schema

```python
{
  'total_steps': int,
  'num_updates': int,
  'episode_rewards': list[float][-100:],
  'episode_lengths': list[int][-100:],
  'win_counts': {'win':int, 'loss':int, 'draw':int},
  'total_episodes': int,
  'config': {...subset for reproducibility...},
  'model_state_dict': policy.state_dict(),
  'optimizer_state_dict': optimizer.state_dict(),
  'opponent_state_dict': <optional>
}
```

### Logging

* CSV at `${LOG_DIR}/${RUN_NAME}/metrics.csv` with headers:
  `step,update,episodes,avg_reward,avg_len,win_rate,draw_rate,policy_loss,value_loss,entropy,clip_frac,approx_kl,rollout_sec,update_sec`
* Print concise summaries, align with CSV entries.
* (Optional) TensorBoard scalars with same keys if `TENSORBOARD=True`.

---

### `python -m training.trainer` flags

* `--seed`, `--max-steps`, `--rollout-length`, `--batch-size`, `--epochs`, `--lr`, `--device`, `--run-name`, `--log-dir`, `--ckpt-dir`, `--use-frozen-opponent`, `--opponent-update-interval`, `--max-kl`, `--clip-eps`, `--entropy-coeff`, `--value-coeff`.
* All flags override config.

### `scripts/run_training.sh`

* Exports `PYTHONHASHSEED`.
* Invokes module with recommended defaults.
* Creates run directory if not exists; echoes run name.

---

### `Dockerfile` (CPU)

* Base: `python:3.10-slim`.
* Install system deps (e.g., `build-essential` if needed, `git`, `procps`).
* Copy `requirements.txt`; `pip install --no-cache-dir -r requirements.txt` with pinned versions.
* Copy source; set `PYTHONUNBUFFERED=1`.
* Default CMD: `bash scripts/run_training.sh`.

### `requirements.txt` (pin)

* `numpy==...`
* `torch==...` (CPU build)
* `tqdm==...`
* `tensorboard==...` (optional)
* `pytest==...`

---

### 10.1 Unit — Policy

* **test_policy_shapes**: encode_obs shape==feature_size; forward returns `(B,4)` and `(B,1)`.
* **test_masking_blocks_illegal**: craft mini board near wall and trail; illegal directions masked (logits ≈ −1e8).
* **test_act_returns_scalars**: action∈[0,3], finite logprob/value.
* **test_evaluate_actions_shapes**: batch of N obs/actions returns tensors of length N; entropy positive.
* **test_kl_guard**: pass NaN old_logprobs; returned KL is finite sentinel.

### 10.2 Unit — Rollout & GAE

* **test_gae_terminal**: single terminal step → return==reward, no bootstrap.
* **test_gae_bootstrap**: non‑terminal last step with mocked `policy.get_value` → return includes bootstrap.
* **test_adv_normalization**: after prepare_batch_data, mean≈0, std≈1.
* **test_mini_batching**: all keys sliced consistently; obs_tensors device==policy.device.

### 10.3 Unit — Trainer

* **test_checkpoint_roundtrip**: save then load → model params identical; counters equal.
* **test_opponent_sync**: after `OPPONENT_UPDATE_INTERVAL` updates, weights equal.
* **test_nan_guard_skip**: forcibly inject NaNs (e.g., set entropy coeff huge) → update step logs warning and continues.

### 10.4 Integration — Short Run

* **test_smoke_training**: run with `MAX_TRAINING_STEPS=4096`, small net; ensure CSV created, checkpoints exist; no exceptions.

> All tests run with `pytest -q` in CPU within 2–3 minutes.

---

## 10) Performance & Memory Budgets

* Policy forward (batch=1): <5ms on typical laptop CPU. Document measurables.
* Model params <5MB; print warning if exceeded.
* Rollout collector uses Python lists; convert to NumPy at end to minimize overhead.

---

## 11) Metrics to Watch (and Alerts)

* **Entropy** too low (<0.2) early → increase `ENTROPY_COEFF`.
* **Approx KL** spikes > `MAX_KL` many epochs → lower LR or clip eps.
* **Clip fraction** consistently > 0.6 → reduce LR or advantages scale issues.
* **Win rate** stagnates + entropy collapses → opponent update interval too short/long.

Add counters for:

* illegal‑mask fallbacks (all‑zero masks → all‑ones).
* skipped updates due to non‑finite loss.

---

## 12) Code Quality & Style

* Type hints for public methods.
* Assert statements for critical shape/dtype invariants.
* Docstrings include shapes and dtypes.
* Logging strings prefixed with module tags: `[Policy]`, `[Rollout]`, `[Trainer]`.

---

### Commit Order (squash‑able but reviewable):

1. Config + utils logging scaffolding.
2. Policy: masking, evaluate_actions, numerics.
3. Rollout: collection, GAE fix, batching, device consistency.
4. Trainer: PPO core + early stop + guards + checkpoints.
5. Logging/CSV + CLI flags.
6. Docker + run script.
7. Tests + README.

### Reviewer Checklist

* [ ] All new flags wired to config.
* [ ] Unit/integration tests pass locally and in Docker.
* [ ] Checkpoint resume works from mid‑run.
* [ ] Opponent sync cadence respected.
* [ ] No CUDA calls; CPU only.
* [ ] No NaN/inf in losses during 2k‑step smoke run.

---

## 14) How to Run (README excerpt)

```bash
# quick smoke run
python -m training.trainer --max-steps 4096 --rollout-length 512 --batch-size 128 \
  --epochs 4 --seed 123 --use-frozen-opponent true --device cpu

# resume from checkpoint
python -m training.trainer --ckpt-path training/checkpoints/checkpoint_step_50000.pt

# inside docker
docker build -t tron-ppo:cpu .
docker run --rm -it -v "$PWD:/app" tron-ppo:cpu
```

Expected outputs:

* Console shows update metrics every ~`LOG_INTERVAL` steps.
* `runs/<RUN_NAME>/metrics.csv` created and growing.
* `training/checkpoints/checkpoint_step_*.pt` written.

---

## 15) Future Work (Out of Scope of This PR)

* Opponent pool with ELO selection.
* Curriculum maps / random board sizes.
* Mixed opponents (heuristics + frozen snapshots).
* Conv encoder for larger boards (if CPU budget allows).
* TensorBoard PR enabling default dashboards.

---

## 16) Acceptance Criteria (Hard Checks)

* [ ] `pytest -q` passes all tests in <3 minutes on CPU.
* [ ] `python -m training.trainer --max-steps 4096` finishes without NaNs.
* [ ] Checkpoint resume reproduces next 1k steps deterministically (same seed & config).
* [ ] Win rate vs. initial frozen opponent improves by ≥5pp over the first 50k steps (non‑scientific acceptance, sanity trend only).

---

## 17) Notes on Edge Cases & Pitfalls

* Mask mismatches between rollout and train → MUST pass `action_masks` to `evaluate_actions` (done).
* GAE boundary: use `dones[t]` weighting; last step uses bootstrap only when `final_done=False`.
* Device drift: always call `.to(config.DEVICE)` on batch prep **and** on encoded obs inside minibatches.
* Logging drift: CSV headers and print keys must match for easy diff/trace.

---

**End of Spec.**
